File: backend/scripts/evaluate.py
#!/usr/bin/env python
"""Score detection pipelines against ground-truth labels.

The parity script asks "do we agree with Ultralytics?". This asks the question
that actually matters: "do we find the loons?". They can diverge — our
preprocessing filters a downscale properly where Ultralytics' aliases it, so
byte-identical inputs were never achievable without taking on OpenCV. What we
need to know is whether the difference costs accuracy.

Reports precision, recall and AP@0.5 for each pipeline over a YOLO-format
split, so a preprocessing or model change can be judged on evidence.

    PYTHONPATH=backend ~/Documents/loonet/.venv-train/bin/python \\
        backend/scripts/evaluate.py --split <val.txt> --weights <best.pt>
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

import numpy as np

BACKEND_ROOT = Path(__file__).resolve().parents[1]
if str(BACKEND_ROOT) not in sys.path:
    sys.path.insert(0, str(BACKEND_ROOT))

from app.detection.detector import Detector  # noqa: E402

IOU_MATCH = 0.5


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--split", type=Path, required=True, help="YOLO split .txt")
    parser.add_argument("--weights", type=Path, help="Ultralytics .pt, to include as a baseline")
    parser.add_argument("--onnx", type=Path, default=BACKEND_ROOT / "models" / "loon_v1.onnx")
    parser.add_argument("--conf", type=float, default=0.001, help="low, so AP sees the full curve")
    parser.add_argument("--iou", type=float, default=0.45)
    args = parser.parse_args()

    images = [Path(line) for line in args.split.read_text().split() if line]
    truth = {p: load_labels(p) for p in images}
    total_objects = sum(len(v) for v in truth.values())
    print(f"{len(images)} images, {total_objects} labelled loons\n")

    detector = Detector(args.onnx)
    runs: dict[str, dict[Path, list[tuple[float, np.ndarray]]]] = {}

    runs["onnx single-pass"] = {
        p: [
            (d.confidence, to_xyxy_pixels(d, w, h))
            for d, w, h in _predictions(detector, p, args, tiling=False)
        ]
        for p in images
    }
    runs["onnx tiled"] = {
        p: [
            (d.confidence, to_xyxy_pixels(d, w, h))
            for d, w, h in _predictions(detector, p, args, tiling=True)
        ]
        for p in images
    }

    if args.weights:
        from ultralytics import YOLO

        model = YOLO(str(args.weights))
        baseline = {}
        for p in images:
            r = model.predict(
                str(p), imgsz=640, conf=args.conf, iou=args.iou, verbose=False, device="cpu"
            )[0]
            baseline[p] = [
                (float(b.conf[0]), np.array(b.xyxy[0].tolist(), dtype=np.float64))
                for b in r.boxes
            ]
        runs["ultralytics baseline"] = baseline

    print(f"{'pipeline':24s} {'P':>7s} {'R':>7s} {'AP@0.5':>8s} {'preds':>7s}")
    print("-" * 56)
    for name in ("ultralytics baseline", "onnx single-pass", "onnx tiled"):
        if name not in runs:
            continue
        p, r, ap, n = score(runs[name], truth, total_objects)
        print(f"{name:24s} {p:7.4f} {r:7.4f} {ap:8.4f} {n:7d}")

    return 0


def _predictions(detector: Detector, path: Path, args, *, tiling: bool):
    outcome = detector.detect(
        path, confidence_threshold=args.conf, iou_threshold=args.iou, tiling=tiling
    )
    return [(d, outcome.image_width, outcome.image_height) for d in outcome.predictions]


def to_xyxy_pixels(prediction, width: int, height: int) -> np.ndarray:
    """Percentage box -> pixel corners, for IoU against pixel-space truth."""
    return np.array(
        [
            prediction.x / 100 * width,
            prediction.y / 100 * height,
            (prediction.x + prediction.width) / 100 * width,
            (prediction.y + prediction.height) / 100 * height,
        ],
        dtype=np.float64,
    )


def load_labels(image_path: Path) -> np.ndarray:
    """Read the YOLO sidecar label file as pixel-space xyxy boxes.

    Dimensions come from the shared loader so that ground truth and
    predictions agree on EXIF orientation; reading ``Image.size`` directly
    would put a rotated photo's labels in a transposed coordinate space.
    """
    from app.detection.loader import read_dimensions

    label_path = Path(str(image_path).replace("/images/", "/labels/")).with_suffix(".txt")
    if not label_path.exists():
        return np.zeros((0, 4))

    width, height = read_dimensions(image_path)

    boxes = []
    for line in label_path.read_text().splitlines():
        parts = line.split()
        if len(parts) < 5:
            continue
        _, cx, cy, w, h = (float(v) for v in parts[:5])
        boxes.append(
            [
                (cx - w / 2) * width,
                (cy - h / 2) * height,
                (cx + w / 2) * width,
                (cy + h / 2) * height,
            ]
        )
    return np.array(boxes, dtype=np.float64) if boxes else np.zeros((0, 4))


def iou_matrix(box: np.ndarray, targets: np.ndarray) -> np.ndarray:
    if targets.size == 0:
        return np.zeros(0)
    x1 = np.maximum(box[0], targets[:, 0])
    y1 = np.maximum(box[1], targets[:, 1])
    x2 = np.minimum(box[2], targets[:, 2])
    y2 = np.minimum(box[3], targets[:, 3])
    inter = (x2 - x1).clip(min=0) * (y2 - y1).clip(min=0)
    area_a = (box[2] - box[0]) * (box[3] - box[1])
    area_b = (targets[:, 2] - targets[:, 0]) * (targets[:, 3] - targets[:, 1])
    union = area_a + area_b - inter
    return np.divide(inter, union, out=np.zeros_like(inter), where=union > 0)


def score(predictions, truth, total_objects: int):
    """Greedy IoU matching, then AP by continuous interpolation of the PR curve."""
    rows = []
    for path, preds in predictions.items():
        targets = truth[path]
        claimed = np.zeros(len(targets), dtype=bool)
        for confidence, box in sorted(preds, key=lambda t: t[0], reverse=True):
            ious = iou_matrix(box, targets)
            hit = -1
            if ious.size:
                order = np.argsort(-ious)
                for idx in order:
                    if ious[idx] < IOU_MATCH:
                        break
                    if not claimed[idx]:
                        claimed[idx] = True
                        hit = int(idx)
                        break
            rows.append((confidence, 1 if hit >= 0 else 0))

    if not rows or total_objects == 0:
        return 0.0, 0.0, 0.0, len(rows)

    rows.sort(key=lambda t: t[0], reverse=True)
    tp = np.cumsum([r[1] for r in rows])
    fp = np.cumsum([1 - r[1] for r in rows])
    recall = tp / total_objects
    precision = tp / np.maximum(tp + fp, 1e-12)

    # Monotone envelope, then area under the step curve.
    envelope = np.maximum.accumulate(precision[::-1])[::-1]
    ap = float(np.sum(np.diff(np.concatenate([[0.0], recall])) * envelope))

    # Operating point at the threshold the app actually ships with.
    at = [r for r in rows if r[0] >= 0.25]
    if at:
        hits = sum(r[1] for r in at)
        return hits / len(at), hits / total_objects, ap, len(rows)
    return 0.0, 0.0, ap, len(rows)


if __name__ == "__main__":
    raise SystemExit(main())

File: desktop/src-tauri/examples/evaluate.rs
//! Score the detector against ground-truth labels: the authoritative accuracy
//! check. Run it after any change to pre- or postprocessing, any model
//! re-export, and before turning tiling on for a new kind of imagery.
//!
//! ```sh
//! cargo run --release --example evaluate -- \
//!     --split ~/Documents/loonet/data/annotated/loonnet_v1/val.txt [--tiled] [--model path.onnx]
//! ```
//!
//! Reports precision and recall at the shipped 0.25 threshold and AP@0.5 over
//! the full curve, for a YOLO-format split: one image path per line, labels in
//! the sibling `labels/` folder as `class cx cy w h`, normalised.

use std::path::{Path, PathBuf};

use gavia_lib::detection::detector::default_thread_count;
use gavia_lib::detection::loader;
use gavia_lib::detection::preprocess::TileSettings;
use gavia_lib::detection::{DetectOptions, Detector};

const IOU_MATCH: f64 = 0.5;
/// Low, so AP sees the whole precision-recall curve.
const CURVE_THRESHOLD: f64 = 0.001;
const SHIPPED_THRESHOLD: f64 = 0.25;

type Box = [f64; 4];

fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let value = |flag: &str| {
        args.iter()
            .position(|a| a == flag)
            .and_then(|i| args.get(i + 1))
            .cloned()
    };
    let Some(split) = value("--split") else {
        eprintln!("usage: evaluate --split <val.txt> [--tiled] [--model <loon.onnx>] [--iou 0.45]");
        std::process::exit(2);
    };
    let model = value("--model").map(PathBuf::from).unwrap_or_else(|| {
        Path::new(env!("CARGO_MANIFEST_DIR")).join("resources/models/loon_v1.onnx")
    });
    let iou = value("--iou").and_then(|v| v.parse().ok()).unwrap_or(0.45);
    let tiled = args.iter().any(|a| a == "--tiled");

    let text = std::fs::read_to_string(&split).unwrap_or_else(|e| panic!("{split}: {e}"));
    let images: Vec<PathBuf> = text.split_whitespace().map(PathBuf::from).collect();
    let truth: Vec<Vec<Box>> = images.iter().map(|p| load_labels(p)).collect();
    let total: usize = truth.iter().map(Vec::len).sum();
    println!("{} images, {total} labelled loons\n", images.len());

    let detector = Detector::load(
        &model,
        Some(default_thread_count()),
        TileSettings::default(),
    )
    .unwrap_or_else(|e| panic!("{e}"));
    let options = DetectOptions {
        confidence_threshold: Some(CURVE_THRESHOLD),
        iou_threshold: Some(iou),
        tiling: Some(tiled),
    };

    let predictions: Vec<Vec<(f64, Box)>> = images
        .iter()
        .map(|path| {
            let outcome = detector
                .detect(&std::fs::read(path).unwrap(), options)
                .unwrap();
            let (w, h) = (
                f64::from(outcome.image_width),
                f64::from(outcome.image_height),
            );
            outcome
                .predictions
                .iter()
                .map(|p| {
                    let corners = [
                        p.x / 100.0 * w,
                        p.y / 100.0 * h,
                        (p.x + p.width) / 100.0 * w,
                        (p.y + p.height) / 100.0 * h,
                    ];
                    (p.confidence, corners)
                })
                .collect()
        })
        .collect();

    let (p, r, ap, n) = score(&predictions, &truth, total);
    let name = if tiled {
        "onnx tiled"
    } else {
        "onnx single-pass"
    };
    println!(
        "{:24} {:>7} {:>7} {:>8} {:>7}",
        "pipeline", "P", "R", "AP@0.5", "preds"
    );
    println!("{}", "-".repeat(56));
    println!("{name:24} {p:7.4} {r:7.4} {ap:8.4} {n:7}");
}

/// YOLO label file as pixel-space corners. Dimensions come from the shared
/// loader so truth and predictions agree on EXIF orientation.
fn load_labels(image: &Path) -> Vec<Box> {
    let label = PathBuf::from(image.to_string_lossy().replace("/images/", "/labels/"))
        .with_extension("txt");
    let Ok(text) = std::fs::read_to_string(label) else {
        return Vec::new();
    };
    let (w, h) = loader::read_dimensions(&std::fs::read(image).unwrap()).unwrap();
    let (w, h) = (f64::from(w), f64::from(h));
    text.lines()
        .filter_map(|line| {
            let v: Vec<f64> = line
                .split_whitespace()
                .filter_map(|t| t.parse().ok())
                .collect();
            (v.len() >= 5).then(|| {
                let (cx, cy, bw, bh) = (v[1], v[2], v[3], v[4]);
                [
                    (cx - bw / 2.0) * w,
                    (cy - bh / 2.0) * h,
                    (cx + bw / 2.0) * w,
                    (cy + bh / 2.0) * h,
                ]
            })
        })
        .collect()
}

fn iou(a: &Box, b: &Box) -> f64 {
    let inter =
        (a[2].min(b[2]) - a[0].max(b[0])).max(0.0) * (a[3].min(b[3]) - a[1].max(b[1])).max(0.0);
    let union = (a[2] - a[0]) * (a[3] - a[1]) + (b[2] - b[0]) * (b[3] - b[1]) - inter;
    if union > 0.0 { inter / union } else { 0.0 }
}

/// Greedy IoU matching per image, then AP by the monotone envelope of the
/// precision-recall curve — the same scoring `evaluate.py` used, so numbers
/// from before and after the Rust port compare directly.
fn score(
    predictions: &[Vec<(f64, Box)>],
    truth: &[Vec<Box>],
    total: usize,
) -> (f64, f64, f64, usize) {
    let mut rows: Vec<(f64, bool)> = Vec::new();
    for (preds, targets) in predictions.iter().zip(truth) {
        let mut claimed = vec![false; targets.len()];
        let mut sorted = preds.clone();
        sorted.sort_by(|a, b| b.0.total_cmp(&a.0));
        for (confidence, candidate) in sorted {
            let mut order: Vec<(usize, f64)> = targets
                .iter()
                .map(|t| iou(&candidate, t))
                .enumerate()
                .collect();
            order.sort_by(|a, b| b.1.total_cmp(&a.1));
            let hit = order
                .into_iter()
                .take_while(|&(_, overlap)| overlap >= IOU_MATCH)
                .find(|&(i, _)| !claimed[i])
                .map(|(i, _)| claimed[i] = true)
                .is_some();
            rows.push((confidence, hit));
        }
    }
    if rows.is_empty() || total == 0 {
        return (0.0, 0.0, 0.0, rows.len());
    }

    rows.sort_by(|a, b| b.0.total_cmp(&a.0));
    let (mut tp, mut fp) = (0.0, 0.0);
    let mut curve: Vec<(f64, f64)> = Vec::with_capacity(rows.len());
    for &(_, hit) in &rows {
        if hit {
            tp += 1.0
        } else {
            fp += 1.0
        }
        curve.push((tp / total as f64, tp / f64::max(tp + fp, 1e-12)));
    }
    let mut envelope = 0.0f64;
    let mut ap = 0.0;
    let mut prev_recall = vec![0.0; curve.len()];
    for i in 1..curve.len() {
        prev_recall[i] = curve[i - 1].0;
    }
    for i in (0..curve.len()).rev() {
        envelope = envelope.max(curve[i].1);
        ap += (curve[i].0 - prev_recall[i]) * envelope;
    }

    let at: Vec<&(f64, bool)> = rows.iter().filter(|r| r.0 >= SHIPPED_THRESHOLD).collect();
    let hits = at.iter().filter(|r| r.1).count() as f64;
    if at.is_empty() {
        return (0.0, 0.0, ap, rows.len());
    }
    (hits / at.len() as f64, hits / total as f64, ap, rows.len())
}
